sympy/solvers/ode/tests/test_systems.py

```diff
@@ -2070,6 +2070,235 @@ def test_linodesolve():
     assert constant_renumber(linodesolve(A, x, type="type4", doit=True), variables=Tuple(*eq).free_symbols) == sol1
 
 
+def test_dsolve_dae():
+    # Test that dsolve can solve a mixture of ODEs and algebraic equations (at
+    # least in simple cases).
+    funcs = [f(x), g(x)]
+
+    eqs123 = [f(x).diff(x) - g(x), g(x)]
+    sol1 = [Eq(f(x), C1), Eq(g(x), 0)]
+    assert dsolve(eqs123, funcs) == sol1
+    assert checksysodesol(eqs123, sol1) == (True, [0, 0])
+
+    ics2 = {f(0):1, g(0): 0}
+    sol2 = [Eq(f(x), 1), Eq(g(x), 0)]
+    assert dsolve(eqs123, funcs, ics=ics2) == sol2
+    assert checksysodesol(eqs123, sol2) == (True, [0, 0])
+
+    # With DAEs it is possible that the initial conditions are inconsistent and
+    # so the existence of a solution to an IVP is not guaranteed. Here the
+    # initial condition for g contradicts the algebraic equation for g:
+    ics3 = {f(0):1, g(0): 1}
+    raises(ValueError, lambda: dsolve(eqs123, funcs, ics=ics3))
+
+    # Purely algebraic equations should be handled at least if linear. Really
+    # something like this should use solve/linsolve et al but simple cases can
+    # be can also be handled here as a base case of DAE systems.
+    eqs4 = [f(x) - g(x), g(x)]
+    sol4 = [Eq(f(x), 0), Eq(g(x), 0)]
+    assert dsolve(eqs4, funcs) == sol4
+    assert checksysodesol(eqs4, sol4) == (True, [0, 0])
+
+    C1f, C2f, C3f, C4f = symbols('C1:5', cls=Function)
+
+    eqs5 = [f(x).diff(x)*g(x)]
+    sol5 = [
+        [Eq(f(x), C1f(x)), Eq(g(x), 0)],
+        [Eq(f(x), C1), Eq(g(x), C2f(x))],
+    ]
```

Comment:
I'm not sure about introducing these functions into the solutions. Maybe it would be better just for some functions to be unspecified in the solutions e.g. the solutions here could be:
```
sol5 = [
   [Eq(g(x), 0)],
   [Eq(f(x), C1)],
]
```
That is more in keeping with what `solve` would do:
```python
In [2]: solve([x*y], [x, y], dict=True)
Out[2]: [{x: 0}, {y: 0}]
```

Reply:
Do we even have the freedom to choose in the tests? Since we are testing for equality with `assert dsolve(eqs5, funcs) == sol5` we need to stick with what output is produced anyways.

Reply:
Now I got it. I would prefer the new output, as it is more informative. I assume this comes in handy especially for larger systems.

Reply:
Saying `g(x) = C2f(x)` is vacuous. It gives no information about `g(x)` in which case it would be better to just leave `g(x)` as `g(x)`.

Reply:
Convinced for now. Maybe at some point I come up with an example where it would be nice to have it, but then I can of course open another issue.